doxygen/man/man3/DarkSilkUnits.3
.TH "DarkSilkUnits" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DarkSilkUnits \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <darksilkunits\&.h>\fP
.PP
Inherits QAbstractListModel\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBUnit\fP { \fBDRKSLK\fP, \fBmDRKSLK\fP, \fBuDRKSLK\fP }"
.br
.ti -1c
.RI "enum \fBSeparatorStyle\fP { \fBseparatorNever\fP, \fBseparatorStandard\fP, \fBseparatorAlways\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBDarkSilkUnits\fP (QObject *parent)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBCAmount\fP \fBmaxMoney\fP ()"
.br
.RI "\fIReturn maximum number of base units (Satoshis) \fP"
.in -1c
.PP
.RI "\fBStatic API\fP"
.br
Unit conversion and formatting 
.PP
.in +1c
.in +1c
.ti -1c
.RI "static QList< \fBUnit\fP > \fBavailableUnits\fP ()"
.br
.RI "\fIGet list of units, for drop-down box\&. \fP"
.ti -1c
.RI "static bool \fBvalid\fP (int unit)"
.br
.RI "\fIIs unit ID valid? \fP"
.ti -1c
.RI "static QString \fBname\fP (int unit)"
.br
.RI "\fIShort name\&. \fP"
.ti -1c
.RI "static QString \fBdescription\fP (int unit)"
.br
.RI "\fILonger description\&. \fP"
.ti -1c
.RI "static qint64 \fBfactor\fP (int unit)"
.br
.RI "\fINumber of Satoshis (1e-8) per unit\&. \fP"
.ti -1c
.RI "static int \fBamountDigits\fP (int unit)"
.br
.RI "\fINumber of amount digits (to represent max number of coins) \fP"
.ti -1c
.RI "static int \fBdecimals\fP (int unit)"
.br
.RI "\fINumber of decimals left\&. \fP"
.ti -1c
.RI "static QString \fBformat\fP (int unit, const \fBCAmount\fP &nIn, bool plussign=false)"
.br
.RI "\fIFormat as string\&. \fP"
.ti -1c
.RI "static QString \fBformatWithUnit\fP (int unit, const \fBCAmount\fP &amount, bool plussign=false)"
.br
.RI "\fIFormat as string (with unit) \fP"
.ti -1c
.RI "static bool \fBparse\fP (int unit, const QString &\fBvalue\fP, \fBCAmount\fP *val_out)"
.br
.RI "\fIParse string to coin amount\&. \fP"
.ti -1c
.RI "static QString \fBfloorWithUnit\fP (int unit, const \fBCAmount\fP &amount, bool plussign=false, \fBSeparatorStyle\fP separators=\fBseparatorStandard\fP)"
.br
.RI "\fIFormat as string (with unit) but floor value up to 'digits' settings\&. \fP"
.ti -1c
.RI "static QString \fBfloorHtmlWithUnit\fP (int unit, const \fBCAmount\fP &amount, bool plussign=false, \fBSeparatorStyle\fP separators=\fBseparatorStandard\fP)"
.br
.ti -1c
.RI "static QString \fBformat\fP (int unit, const \fBCAmount\fP &nIn, bool fPlus, \fBSeparatorStyle\fP separators)"
.br
.in -1c
.in -1c
.SS "AbstractListModel implementation"
 List model for unit drop-down selection box\&. 
.in +1c
.ti -1c
.RI "enum \fBRoleIndex\fP { \fBUnitRole\fP = Qt::UserRole }"
.br
.ti -1c
.RI "int \fBrowCount\fP (const QModelIndex &parent) const "
.br
.ti -1c
.RI "QVariant \fBdata\fP (const QModelIndex &index, int role) const "
.br
.in -1c
.SH "Detailed Description"
.PP 
DarkSilk unit definitions\&. Encapsulates parsing and formatting and serves as list model for drop-down selection boxes\&. 
.PP
Definition at line 33 of file darksilkunits\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBDarkSilkUnits::RoleIndex\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIUnitRole \fP\fP
Unit identifier 
.PP
Definition at line 91 of file darksilkunits\&.h\&.
.SS "enum \fBDarkSilkUnits::SeparatorStyle\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIseparatorNever \fP\fP
.TP
\fB\fIseparatorStandard \fP\fP
.TP
\fB\fIseparatorAlways \fP\fP
.PP
Definition at line 50 of file darksilkunits\&.h\&.
.SS "enum \fBDarkSilkUnits::Unit\fP"
DarkSilk units\&. 
.PP
\fBNote:\fP
.RS 4
Source: https://en.darksilk.it/wiki/Units \&. Please add only sensible ones 
.RE
.PP

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIDRKSLK \fP\fP
.TP
\fB\fImDRKSLK \fP\fP
.TP
\fB\fIuDRKSLK \fP\fP
.PP
Definition at line 43 of file darksilkunits\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "DarkSilkUnits::DarkSilkUnits (QObject * parent)\fC [explicit]\fP"

.PP
Definition at line 12 of file darksilkunits\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "int DarkSilkUnits::amountDigits (int unit)\fC [static]\fP"

.PP
Number of amount digits (to represent max number of coins) 
.PP
Definition at line 73 of file darksilkunits\&.cpp\&.
.SS "QList< \fBDarkSilkUnits::Unit\fP > DarkSilkUnits::availableUnits ()\fC [static]\fP"

.PP
Get list of units, for drop-down box\&. 
.PP
Definition at line 18 of file darksilkunits\&.cpp\&.
.SS "QVariant DarkSilkUnits::data (const QModelIndex & index, int role) const"

.PP
Definition at line 217 of file darksilkunits\&.cpp\&.
.SS "int DarkSilkUnits::decimals (int unit)\fC [static]\fP"

.PP
Number of decimals left\&. 
.PP
Definition at line 84 of file darksilkunits\&.cpp\&.
.SS "QString DarkSilkUnits::description (int unit)\fC [static]\fP"

.PP
Longer description\&. 
.PP
Definition at line 51 of file darksilkunits\&.cpp\&.
.SS "qint64 DarkSilkUnits::factor (int unit)\fC [static]\fP"

.PP
Number of Satoshis (1e-8) per unit\&. 
.PP
Definition at line 62 of file darksilkunits\&.cpp\&.
.SS "QString DarkSilkUnits::floorHtmlWithUnit (int unit, const \fBCAmount\fP & amount, bool plussign = \fCfalse\fP, \fBSeparatorStyle\fP separators = \fC\fBseparatorStandard\fP\fP)\fC [static]\fP"

.PP
Definition at line 139 of file darksilkunits\&.cpp\&.
.SS "QString DarkSilkUnits::floorWithUnit (int unit, const \fBCAmount\fP & amount, bool plussign = \fCfalse\fP, \fBSeparatorStyle\fP separators = \fC\fBseparatorStandard\fP\fP)\fC [static]\fP"

.PP
Format as string (with unit) but floor value up to 'digits' settings\&. 
.PP
Definition at line 128 of file darksilkunits\&.cpp\&.
.SS "QString DarkSilkUnits::format (int unit, const \fBCAmount\fP & nIn, bool plussign = \fCfalse\fP)\fC [static]\fP"

.PP
Format as string\&. 
.PP
Definition at line 95 of file darksilkunits\&.cpp\&.
.SS "QString DarkSilkUnits::format (int unit, const \fBCAmount\fP & nIn, bool fPlus, \fBSeparatorStyle\fP separators)\fC [static]\fP"

.PP
Definition at line 146 of file darksilkunits\&.cpp\&.
.SS "QString DarkSilkUnits::formatWithUnit (int unit, const \fBCAmount\fP & amount, bool plussign = \fCfalse\fP)\fC [static]\fP"

.PP
Format as string (with unit) 
.PP
Definition at line 123 of file darksilkunits\&.cpp\&.
.SS "\fBCAmount\fP DarkSilkUnits::maxMoney ()\fC [static]\fP"

.PP
Return maximum number of base units (Satoshis) 
.PP
Definition at line 237 of file darksilkunits\&.cpp\&.
.SS "QString DarkSilkUnits::name (int unit)\fC [static]\fP"

.PP
Short name\&. 
.PP
Definition at line 40 of file darksilkunits\&.cpp\&.
.SS "bool DarkSilkUnits::parse (int unit, const QString & value, \fBCAmount\fP * val_out)\fC [static]\fP"

.PP
Parse string to coin amount\&. 
.PP
Definition at line 174 of file darksilkunits\&.cpp\&.
.SS "int DarkSilkUnits::rowCount (const QModelIndex & parent) const"

.PP
Definition at line 211 of file darksilkunits\&.cpp\&.
.SS "bool DarkSilkUnits::valid (int unit)\fC [static]\fP"

.PP
Is unit ID valid? 
.PP
Definition at line 27 of file darksilkunits\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for darksilk from the source code\&.


CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix
Stormnode payments start on block 82002
Restore hex to getrawtransaction vout scriptPubkey
static const rename | nActualSpacing code reform
Removal of Keepass until later date that this doesn't cause an issue with Windows builds
Updated dependencies, added Gitian OSX files
Remove unused GetProofOfStakeReward
Faster Retargetting for PoW
Fixed PoS target spacing
Fix payee selectoin on StormnodePayments::ProcessBlock
fix log output / more notCapableReason-s
HasCollateralInputs should check only for at least 1 collateral not 2
Track lastTimeChanged for mnodeman and ask for update only if UpdateNeeded
Added sstx support for free transactions 
Fix for src/qt/stormnodemanager.cpp
RPC Fixes
Fix DoAutomaticDenominating
Few Fixes for Processing Extra Messages
Reversioning and restriction change
Check and remove expired Stormnodes on Client start
Maps in snodeman
bool warning fix
SSEG Update
Show new number of SN's on add/remove in logs | Move output to debug level
sigTime
Temp Fix prior to major changes
SPORK Additions/Changes | SANDSTORM fee removal
prevent ganing sigTime too much on initial ssee
SandStorm/InstantX Functionality | Send with SandStorm on sendcoinsdialog.ui
Fix Filtering for active | add spaces to table-fy console view | add spaces to IP only in full mode
Merged Stormnode blinding
fix compiling for windows
fix console spam
Make stormnodes.dat verification more specific, prevent from overwriting corruped
files with unknown format, give more info in debug.log output
fixed shared key usage
fix even more console spam
rename stormnodes.dat->mncache.dat / fix strings
save nsqCount in mncache.dat
[Qt] add more NULL pointer checks in darksilkgui.cpp
close wallet regardless of "minimize to tray" settings
Added coincontrol sandstorm rounds column
Stormnode Proof of Service Added
Add structure for donating a percentage of ones earnings to another party via the protocol
Fixed snw message issue
fix pubkey/donation conditions
Fixed blinding submission process
increasing ss timeouts
fixed ssee sync
fixed Stormnode payments FindOldestNotInVec

src/qt/darksilkunits.cpp
// Copyright (c) 2011-2016 The Bitcoin developers
// Copyright (c) 2014-2016 Silk Network
// Distributed under the MIT/X11 software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include "darksilkunits.h"
#include "primitives/transaction.h"

#include <QStringList>
#include <QSettings>

DarkSilkUnits::DarkSilkUnits(QObject *parent):
        QAbstractListModel(parent),
        unitlist(availableUnits())
{
}

QList<DarkSilkUnits::Unit> DarkSilkUnits::availableUnits()
{
    QList<DarkSilkUnits::Unit> unitlist;
    unitlist.append(DRKSLK);
    unitlist.append(mDRKSLK);
    unitlist.append(uDRKSLK);
    return unitlist;
}

bool DarkSilkUnits::valid(int unit)
{
    switch(unit)
    {
    case DRKSLK:
    case mDRKSLK:
    case uDRKSLK:
        return true;
    default:
        return false;
    }
}

QString DarkSilkUnits::name(int unit)
{
    switch(unit)
    {
    case DRKSLK: return QString("DRKSLK");
    case mDRKSLK: return QString("mDRKSLK");
    case uDRKSLK: return QString::fromUtf8("μDRKSLK");
    default: return QString("???");
    }
}

QString DarkSilkUnits::description(int unit)
{
    switch(unit)
    {
    case DRKSLK: return QString("DarkSilks");
    case mDRKSLK: return QString("Milli-DarkSilks (1 / 1,000)");
    case uDRKSLK: return QString("Micro-DarkSilkss (1 / 1,000,000)");
    default: return QString("???");
    }
}

qint64 DarkSilkUnits::factor(int unit)
{
    switch(unit)
    {
    case DRKSLK:  return 100000000;
    case mDRKSLK: return 100000;
    case uDRKSLK: return 100;
    default:   return 100000000;
    }
}

int DarkSilkUnits::amountDigits(int unit)
{
    switch(unit)
    {
    case DRKSLK: return 8; // 21,000,000 (# digits, without commas)
    case mDRKSLK: return 11; // 21,000,000,000
    case uDRKSLK: return 14; // 21,000,000,000,000
    default: return 0;
    }
}

int DarkSilkUnits::decimals(int unit)
{
    switch(unit)
    {
    case DRKSLK: return 8;
    case mDRKSLK: return 5;
    case uDRKSLK: return 2;
    default: return 0;
    }
}

QString DarkSilkUnits::format(int unit, const CAmount& nIn, bool fPlus)
{
    // Note: not using straight sprintf here because we do NOT want
    // localized number formatting.
    if(!valid(unit))
        return QString(); // Refuse to format invalid unit
    qint64 n = (qint64)nIn;
    qint64 coin = factor(unit);
    int num_decimals = decimals(unit);
    qint64 n_abs = (n > 0 ? n : -n);
    qint64 quotient = n_abs / coin;
    qint64 remainder = n_abs % coin;
    QString quotient_str = QString::number(quotient);
    QString remainder_str = QString::number(remainder).rightJustified(num_decimals, '0');

    // Right-trim excess zeros after the decimal point
    int nTrim = 0;
    for (int i = remainder_str.size()-1; i>=2 && (remainder_str.at(i) == '0'); --i)
        ++nTrim;
    remainder_str.chop(nTrim);

    if (n < 0)
        quotient_str.insert(0, '-');
    else if (fPlus && n > 0)
        quotient_str.insert(0, '+');
    return quotient_str + QString(".") + remainder_str;
}

QString DarkSilkUnits::formatWithUnit(int unit, const CAmount& amount, bool plussign)
{
    return format(unit, amount, plussign) + QString(" ") + name(unit);
}

QString DarkSilkUnits::floorWithUnit(int unit, const CAmount& amount, bool plussign, SeparatorStyle separators)
{
    QSettings settings;
    int digits = settings.value("digits").toInt();

    QString result = format(unit, amount, plussign, separators);
    if(decimals(unit) > digits) result.chop(decimals(unit) - digits);

    return result + QString(" ") + name(unit);
}

QString DarkSilkUnits::floorHtmlWithUnit(int unit, const CAmount& amount, bool plussign, SeparatorStyle separators)
{
    QString str(floorWithUnit(unit, amount, plussign, separators));
    str.replace(QChar(THIN_SP_CP), QString(THIN_SP_HTML));
    return QString("<span style='white-space: nowrap;'>%1</span>").arg(str);
}

QString DarkSilkUnits::format(int unit, const CAmount& nIn, bool fPlus, SeparatorStyle separators)
{
    // Note: not using straight sprintf here because we do NOT want
    // localized number formatting.
    if(!valid(unit))
        return QString(); // Refuse to format invalid unit
    qint64 n = (qint64)nIn;
    qint64 coin = factor(unit);
    int num_decimals = decimals(unit);
    qint64 n_abs = (n > 0 ? n : -n);
    qint64 quotient = n_abs / coin;
    qint64 remainder = n_abs % coin;
    QString quotient_str = QString::number(quotient);
    QString remainder_str = QString::number(remainder).rightJustified(num_decimals, '0');

    // Right-trim excess zeros after the decimal point
    int nTrim = 0;
    for (int i = remainder_str.size()-1; i>=2 && (remainder_str.at(i) == '0'); --i)
        ++nTrim;
    remainder_str.chop(nTrim);

    if (n < 0)
        quotient_str.insert(0, '-');
    else if (fPlus && n > 0)
        quotient_str.insert(0, '+');
    return quotient_str + QString(".") + remainder_str;
}

bool DarkSilkUnits::parse(int unit, const QString &value, CAmount *val_out)
{
    if(!valid(unit) || value.isEmpty())
        return false; // Refuse to parse invalid unit or empty string
    int num_decimals = decimals(unit);
    QStringList parts = value.split(".");

    if(parts.size() > 2)
    {
        return false; // More than one dot
    }
    QString whole = parts[0];
    QString decimals;

    if(parts.size() > 1)
    {
        decimals = parts[1];
    }
    if(decimals.size() > num_decimals)
    {
        return false; // Exceeds max precision
    }
    bool ok = false;
    QString str = whole + decimals.leftJustified(num_decimals, '0');

    if(str.size() > 18)
    {
        return false; // Longer numbers will exceed 63 bits
    }
    CAmount retvalue = str.toLongLong(&ok);
    if(val_out)
    {
        *val_out = retvalue;
    }
    return ok;
}

int DarkSilkUnits::rowCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return unitlist.size();
}

QVariant DarkSilkUnits::data(const QModelIndex &index, int role) const
{
    int row = index.row();
    if(row >= 0 && row < unitlist.size())
    {
        Unit unit = unitlist.at(row);
        switch(role)
        {
        case Qt::EditRole:
        case Qt::DisplayRole:
            return QVariant(name(unit));
        case Qt::ToolTipRole:
            return QVariant(description(unit));
        case UnitRole:
            return QVariant(static_cast<int>(unit));
        }
    }
    return QVariant();
}

CAmount DarkSilkUnits::maxMoney()
{
    return MAX_MONEY;
}

QString DarkSilkUnits::getAmountColumnTitle(int unit)
{
    QString amountTitle = QObject::tr("Amount");
    if (DarkSilkUnits::valid(unit))
    {
        amountTitle += " ("+DarkSilkUnits::name(unit) + ")";
    }
    return amountTitle;
}

